src/middleware.ts

```diff
@@ -259,35 +262,55 @@ export const persist = <S extends State>(
   }
 
   // rehydrate initial state with existing stored state
-  ;(async () => {
+  ;(() => {
     const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
 
-    try {
-      const storageValue = await storage.getItem(name)
+    let storageValueUnknown = storage.getItem(name)
 
-      if (storageValue) {
-        const deserializedStorageValue = await deserialize(storageValue)
-
-        // if versions mismatch, run migration
-        if (deserializedStorageValue.version !== version) {
-          const migratedState = await migrate?.(
-            deserializedStorageValue.state,
-            deserializedStorageValue.version
-          )
+    if (storageValueUnknown instanceof Promise) {
+      storageValueUnknown
+        .then((storageValue) => storageValue && deserialize(storageValue))
+        .then((deserializedStorageValue) => {
+          if (
+            deserializedStorageValue &&
+            deserializedStorageValue.version !== version
+          ) {
+            return migrate?.(
+              deserializedStorageValue.state,
+              deserializedStorageValue.version
+            )
+          } else {
+            set(deserializedStorageValue.state)
+            return null
+          }
+        })
+        .then((migratedState) => {
           if (migratedState) {
             set(migratedState)
-            await setItem()
+            return setItem()
           }
+        })
+        .then(() => {
+          postRehydrationCallback?.(get(), undefined)
+        })
+        .catch((e) => {
+          postRehydrationCallback?.(undefined, e)
+        })
+    } else {
+      try {
+        let deserializedValueUnknown = deserialize(
+          storageValueUnknown as string
+        )
+        if (deserializedValueUnknown instanceof Promise) {
+          deserializedValueUnknown.then((result) => set(result.state))
```

Comment:
We need to run postRehydrationCallback after this is resolved or rejected.

Reply:
Actually, we lose `migrate` in L300 condition branch.

Hmm, this looks hard. You might want to create some functions and compose them.

Reply:
Ok, thank you I will fix that.